Add com.snipaste.Snipaste

chenzhiwei opened the pull request:
<!-- ⚠️⚠️ Submission pull request MUST be made against the `new-pr` **base branch** ⚠️⚠️  -->

<!-- 💡 Go to the preview tab to click the links below 💡 -->

### Please confirm your submission meets all the criteria

<!-- 💡 Please replace each `[ ]` with `[X]` when the step is complete 💡 -->

- [x] Please describe the application briefly. <!-- insert the description here --> Snipaste is a versatile and user-friendly application designed to enhance productivity through its core features: snipping, pasting, and image annotation.
- [x] The domain used for the application ID is [controlled by the application developer(s)][appid-domain] and the [application id guidelines][appid] are followed.
- [x] I have read and followed all the [Submission requirements][reqs] and the [Submission guide][reqs2].
- [x] I have [built][build] and tested the submission locally.
- [x] I am an author/developer/upstream contributor to the project. If not, I contacted upstream developers about this submission. **Link:** https://github.com/Snipaste/feedback/issues/3242

<!-- ⚠️⚠️ DO NOT modify anything below this line ⚠️⚠️  -->

[appid-domain]: https://docs.flathub.org/docs/for-app-authors/requirements/#control-over-domain-or-repository
[appid]: https://docs.flathub.org/docs/for-app-authors/requirements#application-id
[reqs]: https://docs.flathub.org/docs/for-app-authors/requirements
[reqs2]: https://docs.flathub.org/docs/for-app-authors/submission
[build]: https://docs.flathub.org/docs/for-app-authors/submission/#before-submission

bbhtt commented on com.snipaste.Snipaste.metainfo.xml:
```diff
@@ -0,0 +1,42 @@
+  <url type="homepage">https://www.snipaste.com</url>
+  <url type="help">https://www.snipaste.com/support.html</url>
+  <description>
+    <p>Snipaste is not just a snipping tool. It will boost your productivity if you get used to its 'paste'!</p>
+    <p>NOTE: This wrapper is not verified by, affiliated with, or supported by Snipaste.</p>
```
```suggestion
    <p>NOTE: This wrapper is not verified by, affiliated with, or supported by Snipaste.</p>
    <p>Snipaste is not just a snipping tool. It will boost your productivity if you get used to its 'paste'!</p>
```

Make this the first p tag.

chenzhiwei replied:
fixed

bbhtt commented on com.snipaste.Snipaste.metainfo.xml:
```diff
@@ -0,0 +1,42 @@
+  <launchable type="desktop-id">com.snipaste.Snipaste.desktop</launchable>
+  <url type="homepage">https://www.snipaste.com</url>
+  <url type="help">https://www.snipaste.com/support.html</url>
+  <description>
```
also can you add some more description?

chenzhiwei replied:
added

bbhtt commented on com.snipaste.Snipaste.metainfo.xml:
```diff
@@ -0,0 +1,42 @@
+  </description>
+  <screenshots>
+    <screenshot type="default">
+      <image>https://i.v2ex.co/71Ftp04b.png</image>
```
if this is a temporary image host, better to add these in the PR or they will be lost.

bbhtt replied:
also why are these screenshots on windows?

chenzhiwei replied:
![image](https://github.com/user-attachments/assets/c2df18d0-e4b5-4ff6-b121-9b2fe8c99a3d)

chenzhiwei replied:
The official Snipaste has no other screenshots. https://www.snipaste.com/

bbhtt replied:
Try taking some screenshots on Linux and add them to the PR there. Then update the link to the metainfo.

bbhtt commented on com.snipaste.Snipaste.yaml:
```diff
@@ -0,0 +1,42 @@
+separate-locales: false
+
+finish-args:
+  - --device=all
```
why does it need all?

chenzhiwei replied:
Give me some time to test these args.

chenzhiwei replied:
changed to dri

bbhtt commented on com.snipaste.Snipaste.yaml:
```diff
@@ -0,0 +1,42 @@
+  - --share=network
+
+  - --filesystem=xdg-pictures
+  - --filesystem=xdg-run/pipewire-0
```
`xdg-run/pipewire-0` why is this needed?

bbhtt replied:
also does it work with `:ro`?

chenzhiwei replied:
removed pipwire. The pictures is used to store the screenshot.

bbhtt commented on com.snipaste.Snipaste.yaml:
```diff
@@ -0,0 +1,42 @@
+
+  - --env=XCURSOR_PATH=/run/host/user-share/icons:/run/host/share/icons
+
+  # D-Bus Access
+  - --talk-name=org.freedesktop.Notifications
```
if it uses libnotify for notifications and libnotify 0.8 is present in the appimage this won't be needed.

chenzhiwei replied:
removed this dbus access

bbhtt commented on com.snipaste.Snipaste.yaml:
```diff
@@ -0,0 +1,42 @@
+  - --talk-name=org.freedesktop.Notifications
+
+modules:
+  - name: Snipaste
+    buildsystem: simple
+    build-commands:
```
this needs to be extra-data if redistribution is not allowed.

chenzhiwei replied:
fixed by using extra data

bbhtt commented on com.snipaste.Snipaste.metainfo.xml:
```diff
@@ -0,0 +1,42 @@
+  <developer_name>Snipaste</developer_name>
+  <summary>Snipaste is not just a snipping tool. It will boost your productivity if you get used to its 'paste'!</summary>
+  <metadata_license>CC0-1.0</metadata_license>
+  <launchable type="desktop-id">com.snipaste.Snipaste.desktop</launchable>
```
is redistribution allowed?

chenzhiwei replied:
waiting for official reply

chenzhiwei replied:
Will this block the redistribution?

liulex replied:
I‘m the author of Snipaste. I approve this redistribution. Thank you @chenzhiwei !

bbhtt replied:
thanks. do you want to be added to the repo as a collaborator if/when this is merged?

bbhtt commented:
> - [ ] I am an author/developer/upstream contributor to the project. If not, I contacted upstream developers about this submission. **Link:**

Unless you are the author please contact the developers of the app about the submission.

hfiguiere commented:
> Please describe the application briefly. 

This is missing

hfiguiere commented on com.snipaste.Snipaste.yaml:
```diff
@@ -0,0 +1,42 @@
+    build-commands:
+      - chmod +x Snipaste.AppImage
+      - ./Snipaste.AppImage --appimage-extract
+      - install -Dm755 squashfs-root/usr/bin /app/usr/bin
```
`/app/usr` isn't a prefix and doesn't make sense.

hfiguiere commented on com.snipaste.Snipaste.yaml:
```diff
@@ -0,0 +1,72 @@
+      - type: extra-data
+        only-arches: [x86_64]
+        filename: Snipaste.AppImage
+        url: https://bitbucket.org/liule/snipaste/downloads/Snipaste-2.10.5-x86_64.AppImage
```
why are these coming from a bitbucket and not from `snipaste.com`?

chenzhiwei replied:
Because Snipaste distribute packages from bitbucket.

liulex replied:
Because links from snipaste.com always redirect to the latest version rather than to a specific version.

chenzhiwei commented:
bot, build com.snipaste.Snipaste

hfiguiere commented on com.snipaste.Snipaste.desktop:
```diff
@@ -0,0 +1,15 @@
+X-KDE-StartupNotify=false
+X-KDE-UniqueApplet=true
+X-KDE-DBUS-Restricted-Interfaces=org.kde.KWin.ScreenShot2
+X-AppImage-Version=2.10.5
```
```suggestion
```

chenzhiwei commented on screenshot.png:
Because there is no Linux screenshot, but the features are some.

hfiguiere replied:
you can make one

bbhtt commented on com.snipaste.Snipaste.metainfo.xml:
```diff
@@ -0,0 +1,36 @@
+<component type="desktop-application">
+  <id>com.snipaste.Snipaste</id>
+  <name>Snipaste</name>
+  <project_license>LicenseRef-proprietary=https://www.snipaste.com/index.html</project_license>
```
https://www.snipaste.com/index.html doesn't go to the license.

chenzhiwei replied:
fixed

chenzhiwei commented:
bot, build com.snipaste.Snipaste

chenzhiwei commented:
@bbhtt @hfiguiere could you take a look again?

hfiguiere commented:
> Please describe the application briefly.

This question was never answered.

bbhtt commented:
there are a bunch of unresolved comments.

bbhtt commented on com.snipaste.Snipaste.yaml:
```diff
@@ -0,0 +1,71 @@
+      - type: file
+        path: com.snipaste.Snipaste.metainfo.xml
+
+      - type: extra-data
```
you can switch away from extra-data, according to https://github.com/flathub/flathub/pull/6204#discussion_r1976432485 redistribution is allowed.

chenzhiwei replied:
sure, will do this later.

chenzhiwei commented:
In the latest update:

1. Added the description
2. Updated the screenshot
3. Switched away from extra-data as the Author approved this distribution

Please also add @liulex as collaborator.

But I encountered an issue during the manual testing, please just review and hold on merge.

The issue is: when I run with flatpak I got the message `Qt: Session management error: None of the authentication protocols specified are supported`

@liulex @bbhtt @hfiguiere Do you have any ideas why the subprocess is not running inside flatpak?

Following is running well outside flatpak(the wlhelper is running):
![image](https://github.com/user-attachments/assets/13c14b86-c567-4cf9-a8c1-b426b26e1776)

Following is running in flatpak:
![image](https://github.com/user-attachments/assets/accc0329-0522-405f-be38-a980b27e355f)

bbhtt commented:
bot, build com.snipaste.Snipaste

bbhtt commented:
The Qt error is fine. But this does not seem to work on GNOME wayland.


What environments are supported?

https://github.com/user-attachments/assets/27fde587-7e98-4057-8f18-536a823de318

bbhtt commented:
The icon is a bit blurry

![image](https://github.com/user-attachments/assets/db0e208e-3200-4177-a2f5-89fff9864031)

chenzhiwei commented:
> The Qt error is fine. But this does not seem to work on GNOME wayland.
> 
> What environments are supported?

There should be two processes, the main `Snipaste` and `wlhelper -platform wayland`.

But the `wlhelper` process is not running in the flatpak, I think this is the reason why it does not work.

@liulex Could you give some tips on how does Snipaste starts the wlhelper subprocess?

hfiguiere commented:
if it require something running out of the sanbox, what's the point of the flatpak?

chenzhiwei commented:
> if it require something running out of the sanbox, what's the point of the flatpak?

No, `wlhelper` is inside the flatpak, it is part of Snipaste.

I am trying to figure out why it does not start when running in flatpak.

bbhtt commented:
Also it'd be nice if it showed a dialogue before going to tray. Not every desktop environment has tray support - GNOME doesn't and you need an extension. Without it it will open invisible.

If there is not a dialogue already you can wrap it in `zenity` to provide a dialogue saying:

`this will open in tray, if you are on gnome please install tray extension` etc.

liulex commented:
> Could you give some tips on how does Snipaste starts the wlhelper subprocess?

Something like: @chenzhiwei 
```cpp
m_helperProcess = std::make_unique<QProcess>(parent);
m_helperProcess->start(QApplication::applicationDirPath() + QStringLiteral("/wlhelper"), ...);
```

chenzhiwei commented:
Still investigating:

## Running with Wayland encounter following errors:

```
$ flatpak run --command=bash --socket=wayland --nosocket=x11 com.snipaste.Snipaste
[📦 com.snipaste.Snipaste ~]$ Snipaste
qt.qpa.xcb: could not connect to display
qt.qpa.plugin: Could not load the Qt platform plugin "xcb" in "/app/plugins" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: wayland, xcb.

Aborted (core dumped)
[📦 com.snipaste.Snipaste ~]$ wlhelper -platform wayland
org.freedesktop.DBus.Error.ServiceUnknown
[📦 com.snipaste.Snipaste ~]$
```

## Runninig with x11 encounter following errors:

```
$ flatpak run --command=bash --socket=x11 --nosocket=wayland com.snipaste.Snipaste
[📦 com.snipaste.Snipaste ~]$ Snipaste
Qt: Session management error: Could not open network socket


^Z
[1]+  Stopped                 Snipaste
[📦 com.snipaste.Snipaste ~]$ bg 1
[1]+ Snipaste &
[📦 com.snipaste.Snipaste ~]$ wlhelper -platform wayland
Failed to create wl_display (No such file or directory)
qt.qpa.plugin: Could not load the Qt platform plugin "wayland" in "/app/plugins" even though it was found.
This application failed to start because no Qt platform plugin could be initialized. Reinstalling the application may fix this problem.

Available platform plugins are: wayland, xcb.

Aborted (core dumped)
```

chenzhiwei commented:
> Also it'd be nice if it showed a dialogue before going to tray. Not every desktop environment has tray support - GNOME doesn't and you need an extension. Without it it will open invisible.

@bbhtt  Actually it has a dialog when starting up.

Run following command can start Snipaste correctly, and you will see the dialog(but still can't take screenshot).

This seems bypass the permissions, and I don't know how to find the required permissions.

```
flatpak run --socket=x11 --socket=wayland --socket=session-bus com.snipaste.Snipaste
```

Some findings:

```
$ flatpak run --socket=x11 --socket=wayland --log-session-bus --log-system-bus com.snipaste.Snipaste
C1: -> org.freedesktop.DBus call org.freedesktop.DBus.Hello at /org/freedesktop/DBus
C-65536: -> org.freedesktop.DBus fake wildcarded AddMatch for org.freedesktop.portal
C-65535: -> org.freedesktop.DBus fake AddMatch for org.kde.kwin.Screenshot
C-65534: -> org.freedesktop.DBus fake GetNameOwner for org.kde.kwin.Screenshot
C-65533: -> org.freedesktop.DBus fake wildcarded AddMatch for org.mpris.MediaPlayer2.com.snipaste.Snipaste
C-65532: -> org.freedesktop.DBus fake AddMatch for org.kde.KWin.ScreenShot2
C-65531: -> org.freedesktop.DBus fake GetNameOwner for org.kde.KWin.ScreenShot2
C-65530: -> org.freedesktop.DBus fake wildcarded AddMatch for com.snipaste.Snipaste
C-65529: -> org.freedesktop.DBus fake ListNames
```

```
[2025-03-07 15:17:57.244] [I] Screenshot source: org.kde.KWin.ScreenShot2
[2025-03-07 15:17:57.250] [W] DBus call to screenshot failed: The process is not authorized to take a screenshot
[2025-03-07 15:17:57.250] [W] Null screenshot
```
